ledger-core/conveyor/smachine/exec_adapter.go

```diff
@@ -33,20 +33,41 @@ func (p ExecutionAdapter) GetAdapterID() AdapterID {
 }
 
 func (p ExecutionAdapter) PrepareSync(ctx ExecutionContext, fn AdapterCallFunc) SyncCallRequester {
+	if fn == nil {
+		panic(throw.IllegalValue())
+	}
 	ec := ctx.(*executionContext)
 	return &adapterSyncCallRequest{
 		adapterCallRequest{ctx: ec, fn: fn, adapterID: p.adapterID, isLogging: ec.s.getAdapterLogging(),
 			executor: p.executor, mode: adapterSyncCallContext}}
 }
 
 func (p ExecutionAdapter) PrepareAsync(ctx ExecutionContext, fn AdapterCallFunc) AsyncCallRequester {
+	if fn == nil {
+		panic(throw.IllegalValue())
+	}
 	ec := ctx.(*executionContext)
 	return &adapterCallRequest{ctx: ec, fn: fn, adapterID: p.adapterID, isLogging: ec.s.getAdapterLogging(),
 		executor: p.executor, mode: adapterAsyncCallContext, flags: AutoWakeUp}
 }
 
 func (p ExecutionAdapter) PrepareNotify(ctx ExecutionContext, fn AdapterNotifyFunc) NotifyRequester {
+	if fn == nil {
+		panic(throw.IllegalValue())
+	}
 	ec := ctx.(*executionContext)
 	return &adapterNotifyRequest{ctx: ec, fn: fn, adapterID: p.adapterID, isLogging: ec.s.getAdapterLogging(),
 		executor: p.executor, mode: adapterAsyncCallContext}
 }
+
+func (p ExecutionAdapter) SendFailureNotify(ctx FailureContext, fn AdapterNotifyFunc) {
```

Comment:
split it in two: PrepareFailureNotify and SendFailureNotify. the last one is syntax sugar on top of first. May be I'm wrong and DelayedSend doesn't make sense for FailureNotify.

Reply:
DelayedSend can't be applied here, as FailureContext expects no result, hence there will be only one method - Send()